### PR TITLE
Fix incorrect syntax suggestion with `pub async fn`

### LINKED ISSUE #96555: Incorrect syntax on suggestion to make pub function async with E0277
Given the following code: 
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=0db76b4ce2ff53636126640d313e2b91

```rust
fn main() {
   bar();
}

async fn bar() {
    a_mod::foo().await;
}

mod a_mod {
    pub fn foo() {}
}
```

The current output is:

```
error[E0277]: `()` is not a future
 --> a.rs:6:17
  |
6 |     a_mod::foo().await;
  |     ------------^^^^^^ `()` is not a future
  |     |
  |     this call returns `()`
  |
  = help: the trait `Future` is not implemented for `()`
  = note: () must be a future or must implement `IntoFuture` to be awaited
  = note: required because of the requirements on the impl of `IntoFuture` for `()`
help: remove the `.await`
  |
6 -     a_mod::foo().await;
6 +     a_mod::foo();
  |
help: alternatively, consider making `fn foo` asynchronous
  |
10|     async pub fn foo() {}
  |     +++++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

Ideally the output should look like:

```
...
10|     pub async fn foo() {}
  |         +++++
...
```

Tested on rustc 1.60.0 stable and nightly-2022-04-29

### PR DESCRIPTION
**Pull request by ken-matsui**
This PR closes: https://github.com/rust-lang/rust/issues/96555

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -1082,8 +1082,18 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                             && let Some(hir::Node::Item(hir::Item { span, ident, .. })) =
                                 hir.get_if_local(*def_id)
                         {
+                            let sm = self.tcx.sess.source_map();
+                            let span = sm
+                                .span_extend_to_prev_str(span.shrink_to_hi(), "fn", false, false)
```
Comment (cjgillot) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1087]: Please avoid using this family of function and only use them as a last resort. The source map is not really compatible with incremental compilation, and a possible source of bugs.

The issue is to put the `async` after the `pub`. The `pub`'s span is called `vis_span`. Can it be implemented using it?

Comment (ken-matsui) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs]: @cjgillot 
Thank you so much for your review!

I could implement the same functionality by using `vis_span`, and it's much clearer.
Sorry for having missed the `vis_span`.

**Pull request merged**